src/routers/bookmark_router.py

```
@@ -0,0 +1,140 @@
+        )
+        def list_bookmarks(
+            user: KeycloakUser = Depends(get_user_or_raise), session: Session = Depends(get_session)
+        ) -> List[BookmarkRead]:
+            bookmarks = session.exec(
+                select(Bookmark).where(Bookmark.user_identifier == user._subject_identifier)
+            ).all()
+            return [
+                BookmarkRead(
+                    resource_identifier=bookmark.resource_identifier,
+                    created_at=bookmark.created_at.isoformat(),
+                )
+                for bookmark in bookmarks
+            ]
+
+        @router.post(
+            path,
+            tags=["Bookmarks"],
+            response_model=BookmarkRead,
+            description="Add a bookmark to an asset for a logged-in user.",
+            status_code=HTTPStatus.OK,
+        )
+        def create_bookmark(
+            bookmark: BookmarkCreate,
+            user: KeycloakUser = Depends(get_user_or_raise),
+            session: Session = Depends(get_session),
+        ) -> BookmarkRead:
+            # # Check if the resource exists
+            if not resource_identifier_exists_in_database(bookmark.resource_identifier, session):
+                raise HTTPException(
+                    status_code=HTTPStatus.NOT_FOUND,
+                    detail=f"Resource {bookmark.resource_identifier} does not exist.",
+                )
+
+            # Prevent duplicate bookmarks
+            existing = session.exec(
+                select(Bookmark).where(
+                    Bookmark.user_identifier == user._subject_identifier,
+                    Bookmark.resource_identifier == bookmark.resource_identifier,
+                )
+            ).first()
+            if existing:
+                raise HTTPException(
+                    status_code=HTTPStatus.CONFLICT,
+                    detail=f"Bookmark already exists for this resource identifier {bookmark.resource_identifier}",
+                )
```

I'm actually thinking it might be better design to simply return the existing bookmark.
When a user (or service) uses this endpoint, all they probably care about is that the bookmark is set for the user. If it already exists, that is still the case. The user would probably _always_ ignore this error (but now needs to add a workaround to deal with the different response).